src/t_hash.c

```diff
@@ -457,10 +463,72 @@ robj *hashTypeLookupWriteOrCreate(client *c, robj *key) {
 }
 
 void hashTypeConvertZiplist(robj *o, int enc) {
+    unsigned char *p, *val;
+    unsigned int vlen;
+    long long lval;
+    char longstr[32] = {0};
+
     serverAssert(o->encoding == OBJ_ENCODING_ZIPLIST);
 
     if (enc == OBJ_ENCODING_ZIPLIST) {
         /* Nothing to do... */
+    } else if (enc == OBJ_ENCODING_LISTPACK) {
+        unsigned char *lp; 
+
+        /* Use ziplist's size to pre-allocate listpack,
+         * avoid realloc when lpAppend. */
+        lp = lpNew(ziplistBlobLen(o->ptr)); 
+        p = ziplistIndex(o->ptr, 0);
+        while (ziplistGet(p, &val, &vlen, &lval)) {
+            if (!val) {
+                vlen = ll2string(longstr, sizeof(longstr), lval);
```

Comment:
there is some silly overhead here converting a long to string (due to the interface of `lpAppend`), so that most likely `lpAppend` will convert it back to a long.
maybe we should consider improving that, not necessarily now..

Reply:
This problem is also present in t_stream(`lpAppendInteger`), which I will optimize after this pr.